moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+              qualifierAnnotations == returnTypeAnnotations
+            ) moshi.nextAdapter(factory, parameterTypes[0], qualifierAnnotations)
+            else moshi.adapter(parameterTypes[0], qualifierAnnotations)
+          }
+
+          override fun fromJson(moshi: Moshi, reader: JsonReader): Any? {
+            val intermediate = delegate!!.fromJson(reader)
+            return invoke(intermediate)
+          }
+        }
+      } else {
+        throw IllegalArgumentException(
+          """Unexpected signature for $method.
+@FromJson method signatures may have one of the following structures:
+    <any access modifier> R fromJson(JsonReader jsonReader) throws <any>;
+    <any access modifier> R fromJson(JsonReader jsonReader, JsonAdapter<any> delegate, <any more delegates>) throws <any>;
+    <any access modifier> R fromJson(T value) throws <any>;
+"""
+        )
+      }
+    }
+
+    /** Returns the matching adapter method from the list.  */
+    private fun get(
+      adapterMethods: List<AdapterMethod>,
+      type: Type,
+      annotations: Set<Annotation>
+    ): AdapterMethod? {
+      var i = 0
+      val size = adapterMethods.size
+      while (i < size) {
+        val adapterMethod = adapterMethods[i]
+        if (Types.equals(adapterMethod.type, type) && adapterMethod.annotations == annotations) {
+          return adapterMethod
+        }
+        i++
+      }
+      return null
+    }
+  }
+
+  internal abstract class AdapterMethod(
+    type: Type,
+    val annotations: Set<Annotation>,
+    val adapter: Any,
+    val method: Method,
+    parameterCount: Int,
+    private val adaptersOffset: Int,
+    val nullable: Boolean
+  ) {
+    val type = type.canonicalize()
+    private val jsonAdapters: Array<JsonAdapter<*>?> = arrayOfNulls(parameterCount - adaptersOffset)
+
+    open fun bind(moshi: Moshi, factory: JsonAdapter.Factory) {
+      if (jsonAdapters.isNotEmpty()) {
+        val parameterTypes = method.genericParameterTypes
+        val parameterAnnotations = method.parameterAnnotations
+        var i = adaptersOffset
+        val size = parameterTypes.size
+        while (i < size) {
+          val type = (parameterTypes[i] as ParameterizedType).actualTypeArguments[0]
+          val jsonAnnotations = parameterAnnotations[i].jsonAnnotations
+          jsonAdapters[i - adaptersOffset] =
+            if (Types.equals(this.type, type) && annotations == jsonAnnotations) {
+              moshi.nextAdapter(
+                factory, type, jsonAnnotations
+              )
+            } else {
+              moshi.adapter<Any>(type, jsonAnnotations)
+            }
+          i++
+        }
+      }
+    }
+
+    open fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?): Unit = throw AssertionError()
+
+    open fun fromJson(moshi: Moshi, reader: JsonReader): Any? = throw AssertionError()
+
+    /** Invoke the method with one fixed argument, plus any number of JSON adapter arguments.  */
```

Comment:
Please fix all these single-line docs with two spaces at the end

```suggestion
    /** Invoke the method with one fixed argument, plus any number of JSON adapter arguments. */
```

Reply:
https://github.com/square/moshi/pull/1467/commits/3e6e9bda3372997dff18ac5d1567f6b80707074e#diff-e8a15c8ca843b45b53cf5d1b2097710c70dd4b5a9399c074e7fa2b379a772d47R346